### PR TITLE
dash/editmode-event

### PR DESCRIPTION
**Pull request by sebastianbochan**
Added events to the EditMode state.

### CODE REVIEW

File: ts/Dashboards/EditMode/EditMode.ts [hunk 1]
```diff
@@ -558,6 +560,8 @@ class EditMode {
 
         editMode.active = true;
         editMode.isContextDetectionActive = true;
+
+        eventsOptions?.activate.call(editMode);
```
Comment (Vazonik) [inline ts/Dashboards/EditMode/EditMode.ts:564]: When enabling edit mode, an error is thrown on this line and edit mode does not activate. `eventsOptions.activate` seems to be undefined; adding optional chaining to `activate` does not solve the problem.
image

**Review by Vazonik — CHANGES_REQUESTED** [review]
For some reason, this feature doesn't work in my case, switching to edit mode throws an error.

File: ts/Dashboards/EditMode/EditMode.ts (hunk 1 above)
Comment (sebastianbochan) [reply ts/Dashboards/EditMode/EditMode.ts:564]: By mistake, I set `activated` instead of `activate` in the demo.

Thanks!

Comment (Vazonik) [inline ts/Dashboards/EditMode/EditMode.ts:564]:
```suggestion
        eventsOptions?.activate?.call(editMode);
```

File: ts/Dashboards/EditMode/EditMode.ts [hunk 2]
```diff
@@ -602,6 +607,8 @@ class EditMode {
 
         this.editCellContext = void 0;
         this.potentialCellContext = void 0;
+
+        eventsOptions?.deactivate.call(editMode);
```
Comment (Vazonik) [inline ts/Dashboards/EditMode/EditMode.ts:611]:
```suggestion
        eventsOptions?.deactivate?.call(editMode);
```

File: ts/Dashboards/EditMode/EditMode.ts [hunk 3]
```diff
@@ -1061,6 +1068,14 @@ namespace EditMode {
...
          */
         tools?: Tools;
+         /**
+         * Events attached to the component : `activated`, `deactivated`.
+         *
+         * Try it:
+         *
+         * {@link https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/dashboards/edit-mode/events/ | Activated / Deactivated event }
+         */
+         events?: Record<string, Function>;
```
Comment (Vazonik) [inline ts/Dashboards/EditMode/EditMode.ts:1078]: + Too many spaces here. :)
```suggestion
        /**
        * Events attached to the component : `activated`, `deactivated`.
        *
        * Try it:
        *
        * {@link https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/dashboards/edit-mode/events/ | Activated / Deactivated event }
        */
        events?: Record<string, Function>;
```

**Review by Vazonik — CHANGES_REQUESTED** [review]
Thanks, now it works! One more thing: Options can be unused, in which case they are undefined and an error is thrown while toggling the edit mode. We should check if they exist before calling callbacks. [demo](https://jsfiddle.net/BlackLabel/vknxm7cs/)

**Review by Vazonik — APPROVED** [review]

File: samples/dashboards/edit-mode/events/demo.html [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<script src="https://code.highcharts.com/dashboards/dashboards.js"></script>
+<script src="https://code.highcharts.com/highcharts.js"></script>
+<script src="https://code.highcharts.com/dashboards/datagrid.js"></script>
```
Comment (karolkolodziej) [inline samples/dashboards/edit-mode/events/demo.html:3]: It isn't used.

File: samples/dashboards/edit-mode/events/demo.css [hunk 5]
```diff
@@ -0,0 +1,11 @@
+@import url("https://code.highcharts.com/css/highcharts.css");
+@import url("https://code.highcharts.com/css/datagrid.css");
+@import url("https://code.highcharts.com/css/dashboards.css");
+
+#csv {
```
Comment (karolkolodziej) [inline samples/dashboards/edit-mode/events/demo.css:5]: You don't have csv

File: samples/dashboards/edit-mode/events/demo.css [hunk 6]
```diff
@@ -0,0 +1,11 @@
+@import url("https://code.highcharts.com/css/highcharts.css");
+@import url("https://code.highcharts.com/css/datagrid.css");
```
Comment (karolkolodziej) [inline samples/dashboards/edit-mode/events/demo.css:2]: not nececary

**Review by karolkolodziej — CHANGES_REQUESTED** [review]
Test is missing

**Review by karolkolodziej — APPROVED** [review]

File: ts/Dashboards/EditMode/EditMode.ts [hunk 7]
```diff
@@ -558,6 +560,8 @@ class EditMode {
 
         editMode.active = true;
         editMode.isContextDetectionActive = true;
+
+        eventsOptions?.activate?.call(editMode);
```
Comment (pawellysy) [inline ts/Dashboards/EditMode/EditMode.ts:564]: I'd do that with `addEvent`  / `fireEvent`. If users will like to listen to that event from out of the API, there should be such way. So here, should just be `fireEvent` called, and in the constructor the event should be registered.

**Review by pawellysy — CHANGES_REQUESTED** [review]

File: ts/Dashboards/EditMode/EditMode.ts (hunk 7 above)
Comment (sebastianbochan) [reply ts/Dashboards/EditMode/EditMode.ts:564]: My first thought was the same, but I was wondering if adding the event to the editmode object makes sense. Im not convinced.

Comment (karolkolodziej) [reply ts/Dashboards/EditMode/EditMode.ts:564]: I would certainly unify them. As I look at the first one `mount` it's done by `fireEvent` so we should stick with that 😉

Comment (pawellysy) [reply ts/Dashboards/EditMode/EditMode.ts:564]: I'd add the event to the Board class.

**Review by pawellysy — APPROVED** [review]

**Pull request merged**